### DISCUSSION TITLE
Defining preferred slope values

### DISCUSSION THREAD

**Discussion by majkeldzordan** [NightscoutFoundation/xDrip#2028] [Ideas]
Is it possible (and if not - implement it) to specify a preferred range of slope values? Xdrip sets absurd slopes like 0.5 or 1.5, and then the results are either very low on low glucose or very high on high. It would be good if the user could specify the trading limits - for example 0.7-1.1. I know about the engineering mode, I use it, but many users do not have this knowledge and manually setting the slope is too much for them.

**Comment by Navid200** [comment]
But, users just need to calibrate frequently enough to get reasonable accuracy of both slope and intercept.
The reason Engineering mode capability is added is for experienced users who may want to override calibrations.

If this request is for inexperienced users, the best thing for them to do is to calibrate as often as the manufacturer recommends and they will be good to go.

**Comment by majkeldzordan** [comment]
Ok. Then calibrations should be less mathematical and more practical. Or, I don't know, maybe at least the user should get a warning: hey! your slope has been set to 0.5 which means that if your glucose levels are high your readings will be very low! 

Not a day goes by without someone in the groups complaining that xdrip is either very low or very high. And there is always a common denominator: very high or very low slope.

**Reply by Navid200** [reply]
I agree that the slope should not be set too high or too low.  But, what is too high and what is too low?

The reason we need to calibrate is not just the variation in sensors.  It also has to do with the difference in people.  What slope may be too high for one person may not be too high for someone else.
So, we need to have statistics so that we can judge what slopes never happen and based on that, set the too high and too low limits.
I hope those already exist.  I thought we had those already.

If you are asking us to narrow down the range, you should be careful because you may be moving some people out of the range and that won't be good.

No matter what, if people calibrate frequently enough, their readings should become accurate.  But, I know there are individuals who like to calibrate only three times and stop calibrating.  And the scary part of that is that they go to forums and guide others to do the same.
The sensors change as they remain under our skin.  You need to keep calibrating.  If you are experienced and decide that the change over time is negligible for you, that's your choice.

Limiting the range has to be done very carefully only if necessary in my opinion.

**Reply by majkeldzordan** [reply]
In my opinion, too low a slope starts below 0.7 and too high above 1.2.
It would really be ok to set lower / higher values ONLY if the user calibrates at very high and very low glucose levels. Then we have a line that intersects two points over a long distance. Meanwhile, I see very often that by setting the initial calibration to one slope value, you set 1.3. Absurd. On what basis? Even purely theoretically, the slope should then be set to 1 (and the calibration is done only on the basis of incrept). Today, the freshly activated sensor after calibration had s = 1.4 i = -90. It can't work well. Until I knew the engineering mode, it caused me a lot of frustration and unnecessary stinging.

**Reply by psonnera** [reply]
Libre sensors present a large variability in raw answer either during first 24 hours or mid sensor lifespan. It is very difficult to keep the pace and many users wrongly calibrate at dropping or rising BG making this model wrong. Usually they end up with no data display and believe it's a bridge issue. Resetting cals fixes it. I believe it would be better to refuse calibrations based on background BG variation over the last 15 minutes and when trying to calibrate far outside normal range. It would be also cool to have persistent messages when a calibration is either nonconsistent or drives xDrip+ in safety mode. I myself prefer to keep a slope of 1 and make sure the system is super reliable at 100 and below.

**Reply by psonnera** [reply]
As for slope real data it's very difficult to find a compromise between what users want and what is safe. I'm not sure there is a real dataset study for what the usual Libre slope range is.

**Comment by keencave** [comment]
The slope presets defined in xDrip+ were the result of a long process in the past. There were a lot of a lot of discussions about the limits. Every time when trying to reduce the span we got a lot of complaints that is is not reflecting the practical situation for some of the users. They have somehow strange limits which differs a lot to the average of the users. Lets accept this as a fact. I only count the complaints coming from experienced users here, not from FB. Your diabetes may vary.
Having said this I recommend to leave the limits untouched. 
But it may be a good idea to display a warning message if the user enters a region near the limits.

**Comment by ANDREAPIPPI** [comment]
Hi.
I don’t know if I can ask this here, if not I apologize and will delete this message.
Could someone please explain me/us what the values of "slope" and "intercept" mean exactly?
You are speaking here of numbers that are different from what I see in my graph, so it’s impossible for me to follow what you are discussing.
E.g.: today’s “slope” in my graph is 0,73, “intercept” is 17,13, my sensor is calibrated twice a day when the glycaemic value is constant, and xDrip+ values are very consistent with the BG ones.
How are them calculated?
Are they OK or not?
Why they should not be out of a range (and this is not yet well defined yet, if I understand correctly)?
I appreciate very much your effort to make xdrip+ as accurate as possible, but please remember that the final users are people like me, most of us with little or very little knowledge of many of the stuff related to our problem, and that try to learn more and more from these discussions.
So…. help us, please. 
Andrea    😊

**Reply by psonnera** [reply]
displayed bg = raw bg x slope + intercept. slope and intercept are calculated as a linear regression of your calibration points vs raw bg at calibration time + 10 minutes. AFAIK. slope directly affect the resulting sensitivity i.e. high slope high sensitivity and possible overestimation, low slope low sensitivity i.e. possible underestimate. intercept is lowest possible measurable BG and should be below 40 mg/dl.

**Reply by Navid200** [reply]
@ANDREAPIPPI This, discussions, is the best place for asking any questions.

The glucose in the interstitial fluid, what's under the skin, is related to the blood glucose.  In other words, if our blood glucose goes up, the glucose in our interstitial fluid glucose goes up too.  Of course, there is a delay.  But, that's a different important matter.

The CGMs measure the glucose in the interstitial fluid.  But, the main objective is to control the blood glucose.  So, we need a way to calculate what the blood glucose must be if we know what the interstitial fluid glucose is.  The relationship between the two is close to a linear relationship.  The real relationship may not be perfectly linear.  So, if we wanted to have a very accurate system, we would need to have an even more complicated calibration system.

In it's simplest form, we assume that the relationship is linear.  A linear relationship can be formulated with two constants.  Those two are called the slope of the line, and the offset or intercept of the line.
When you look at your calibration graph, the horizontal axis represents the interstitial fluid glucose and vertical axis represents the blood glucose.  

Having the calibration formula in hand, we can calculate what our blood glucose must be if we only know what our interstitial fluid glucose is even if we have never measured our blood glucose exactly at that same point on the graph.  But, for this to be accurate, we need to ensure that the line is valid and accurate.
The way to accomplish that is to perform more and more calibrations.  
But, calibrations should only be done when the trend is flat.  This is because if the blood glucose is changing, because of the delay in the interstitial fluid glucose change with respect to the blood glucose change, we will cause an error.
As long as you only calibrate when the trend is flat, the more calibrations you perform, the more accurate the calibration will become.
But, this is only correct if the assumption that the relationship is linear was a valid assumption.  There are sensors that show nonlinearity outside a range.  So, you hear people say "never calibrate outside the range".

Now, this puts un in a weird situation:
We start a new sensor and we perform a calibration.  At this point, we have only one point.  How do we create a linear relationship with only one point?  We cannot.  You need at least two points in order to have a line.
This is where it becomes very important to have knowledge of the statistics.  
Statistically speaking, the slope is around 1.0.  Now, people are going to say, no, my slope is always at 0.9.  
This is another part of the problem, the values are not the same for everyone.
That's why I said "statistically".
 When we perform the first calibration, xDrip uses the calibration to figure out what the intercept must be assuming the slope is 1.0.  This value of 1 may be wrong for some people more than for others.
And the value may not be 1.0.  I have to look at the code to know exactly what it is.

Do you think this is complicated?  But, there is more.
Our blood glucose meters are not perfect.  If you use your blood glucose meter to measure your blood exactly at the same time twice, you may not get exactly the same values!  This is because there is random variation in the blood glucose measurements.  That's why most CGMs asks for two initial calibrations instead of just one.  The two calibrations are meant to reduce the impact of the random error of the blood glucose meter.

Unfortunately, some people get lost in trying to achieve perfection.  There is always going to be some error.  We can never 0 out the error.  But, what we need to accomplish is reasonable accuracy.
Another error many fall victim to is assuming their blood glucose meter is perfect.  Your blood glucose meter can be 15% in error and it will still pass the accuracy requirement to be sold in the pharmacy.

Do you think that I am painting a hopeless scenario?
I would like to explain that what matters is the trend and reasonable accuracy and how to accomplish it.
Every time you go to the lab to draw blood and have it tested, do a test with your blood glucose meter at the same time and record the result.
When you get your lab results, compare it to your blood glucose meter measurement you recorded.
Do this every time.  After a few times, if you see a pattern, you can explain this to your doctor and seek advice.
I did this and demonstrated to my doctor that my blood glucose meter was consistently measuring low.  She agreed that I can add a correction to what my blood glucose meter shows.  Please note that I am not giving you medical advice.  I am suggesting to you to talk to your doctor about this.
And calibrate as often as the manufacturer of your CGM recommends.  Don't listen to people who brag that they only calibrate 3 times and get perfect results.

Everything I have stated is my opinion and is not an official statement.

**Reply by ANDREAPIPPI** [reply]
Hi Navid.
Thanks a lot from me and surely from a many other people like me, confused by technical problems like this one.
Now I have a clearer vision of the matter, and have some of my previous understandings confirmed, and some to be modified in my mind.
Now I have more clear what is the reason why you are discussing on "limits yes/limits no", and how difficult could be any decision.
Thank you omce again, very appreciated, as always.
Andrea   😊

**Reply by Navid200** [reply]
You're welcome.

I read my post and realized that I completely left out the impact of time.
I would like to add if your calibrations are perfectly accurate today, there is no guarantee they will still be accurate in 2-3 days from now.  Our immune system considers the filament inserted under our skin as a foreign object that must be fought.  The impact of our immune system on the sensor could reduce its sensitivity (slope).  Another reason why the manufacturer recommends a minimum number of ongoing calibrations.

So, the calibrations keep the slope and intercept valid over the sensor manufacturing production spread, over the variation of human body from person to person, and over time as the sensor remains inserted under our skin.

**Reply by majkeldzordan** [reply]
@Navid200 Unfortunately, in our case, even though we always calibrated on stable glucose, the calibrations did not work well. Only when I started to set s = 0.75 and i = 0, I have practically 100% accuracy every time I check my blood with a meter. I wonder if I am such an exception?

How can I reset these previous data (except "stop sensor") so that xdrip doesn't set me high slope? It is very possible that it is based on data from the beginning of the disease, when we were not yet aware of how soft works.

I used to test DiaBox (bubble mini app) and there is a restriction there - you will not perform calibration if the last 3 readings did not differ much. Maybe you should think about a similar lock in xdrip?

**Reply by Navid200** [reply]
> How can I reset these previous data (except "stop sensor")

You can reset calibrations without stopping sensor.
![Screenshot_20220306-115411](https://user-images.githubusercontent.com/51497406/156933331-fea78fb3-790b-4899-8891-8bf1875b0af1.png)  
  
  <br/>  
<br/>

> Maybe you should think about a similar lock in xdrip?

xDrip is a tool.  It requires some knowledge to use.
If we assume that the user knows everything and keep adding sophisticated capabilities, we could confuse the inexperienced.
On the other hand, if we assume the user knows nothing and keep adding safety measures, it will become a very complicated code to maintain.

Just look at this thread.  It is about a range that exists in xDrip that is being questioned.
If we add another lock, someone is going to open a thread and say why did you add this lock?  It doesn't let me calibrate.
And of course, you are going to suggest to add a setting so that the user can disable it if they don't want it.

And then, we have these threads where people say there are too many settings, which I agree with.

The fact is that there is no way to make everyone happy.

**Reply by majkeldzordan** [reply]
> You can reset calibrations without stopping sensor.

Hmm, I thought that there are even more archival data (from many previous sensors). Because why the first calibration sets me incrept e.g. 1.1 and not 1.0?

> It doesn't let me calibrate.

But there are also threads in which people block the application because they introduce calibration that differs significantly from the source data :-) So maybe it is worth creating, in addition to the current default and engineering mode, an "easy" mode? It could be selected when the application is launched for the first time and possibly changed afterwards.

I know that I can only dream. I just think aloud. xDrip is a powerful application with many possibilities. In my opinion, by allowing or without informing the user to set slope 0.5 or 1.5 you make them have a lot of frustration (with too much slope) or optimism (with low slope). Many months had passed before I started to understand it, and most people (looking at fb groups) still don't understand it and they probably never will. I added this suggestion here because I think it would be practical and helpful. So much for me, I managed through the engineering mode.

**Reply by Navid200** [reply]
The engineering mode lets you directly dictate to xDrip the slope and the intercept.  That is an extremely dangerous power in the hands of someone who doesn't know what they are doing.  That's why it is in engineering mode.  Engineering mode is not meant to be used for the day to day use.  It is only meant for testing.

I am not the right person to answer why your intercept is set to a specific value when you start.  It is somewhere in the code and thanks to so many different settings and capabilities added to xDrip, it is very difficult to look at the code and try to figure out exactly what it does unless you are the one who added the capability.  Even then, hopefully, you remember!  
So, that's why when someone asks for a new setting, I personally always resist unless it is absolutely necessary.  

But, I don't believe any calibration constant is inherited from a previous sensors.

**Reply by keencave** [reply]
Maybe this could help: https://androidaps.readthedocs.io/en/latest/Hardware/Libre2.html
Please refer to the "best practice" rules at the end of the text for the Libre sensors. This summarize the community results for  calibration of the libre sensors the last years. For Dexcom you have slightly other rules.
Most important is to have your measurements done at different glucose levels. It is obvious that too much calibrations around the average BG level would act like an tilt aixs for the calibration. This is a common error for newbies. More measurments dont result in more accuracy.

**Reply by Navid200** [reply]
More correct calibrations (not when the trend is highly slanted) results in more accuracy due to removing the time effect.

If someone only calibrates around one value but uses the CGM at much higher (or lower) values for dosage decisions, the reason for the serious inaccuracy at high (or low) values is not that he calibrated frequently.  The reason is that he used the CGM for dosage decisions in a range that he never calibrated in.

I agree that one should calibrate in a wider range if/when possible.  
I disagree that one should avoid frequent calibrations as a blanket statement.

**Reply by Navid200** [reply]
If I have to choose between the following two scenarios, I will choose 1.

1- Accuracy at 6.0 and slight inaccuracy at 2.8 and 14 
2- Slight accuracy at all while overall accuracy lacks due to ignoring the impact of immune system on the sensor behavior.

If my CGM tells me that I am at 2.8 but in reality I am at 3.5, I still need to eat.  Makes no difference.
If my CGM tells me I am at 14 but in reality, I am at 13, I need insulin.  Makes no difference.  

But, if my CGM tells me that I am in range 95% of the time with an average of 5.8 but in reality I am not and the average is really 6.2, because I have ignored to calibrate often enough, I would not be happy regardless of how accurate the slope may be.

**Comment by psonnera** [comment]
One of the reasons I recommend calibrating first in the lower normal range is because when I see this screenshot in Facebook, my immediate thought is that a slope of 1 most probably drove intercept above 40 and the owner complains there's a no data in xDrip+. On the left comment screenshot, on the right my mind on it. First calibration in the lower normal range usually prevents this issue. Then once in a stable higher range, add another calibration to define slope. But until you have two stable BG calibrations you can't guess what the slope will be.
![image](https://user-images.githubusercontent.com/33809361/156938832-b515e86a-6f3d-4506-ae75-3e37dde0a12b.png)

**Reply by psonnera** [reply]
On the other hand, when adding a calibration point with non-fixed slope without being very far away of the first calibration might drive slope to a creative value. This is why tighter boundaries are better.

![image](https://user-images.githubusercontent.com/33809361/156939521-7c5c8468-6d8b-4886-bff2-26b310c4dbb2.png)

**Comment by tzachi-dar** [comment]
A few thoughts about calibrations...

The TLDR: Don't calibrate, have a system with no calibration like libre and the g6 do.

There are probably two things that needs to be done in order to have a better calibration system:

1) There are big differences between one person to the other. The system should learn the history of a person and offer better calibrations for him. For example there are people who have a typical slope of 1.05 and others have a slope of 0.7. The system should take that into account. For some people on the last days the slope gets bigger.

2) When calculating slopes and intercepts, a mistake in the intercept (especially an intercept of 30 instead of -20) is fatal. I know of a kid who lost his conscience because of this mistake. This can happen because high points influence the intercept. A good calibration algorithm should use the lows to calculate the intercept and use the highs to calculate the slope. 
A positive intercept for a person should only be allowed if that person is known to have positive intercepts. (see 1).

Since:
1) Most people don't understand what calibration is.
2) Most people don't want to calibrate.
3) High and low means different things for different people (for example, some people reach 200 for a few hours a day, and others never go over 140).

People who understand what calibration, slope and intercepts are don't need the algorithm. They can set the values correctly.

Doing 1,2 requires a lot of work and a lot of data from many people. I find it hard to believe it is feasible.

So, hopefully https://github.com/NightscoutFoundation/xDrip/pull/1887 will solve the issue for most people (at least with libre).

**Comment by ChanceHarrison** [comment]
Does the talk here about intercepts, slope, and calibrations apply to G6 transmitters that are in native mode (use the transmitter's internal algorithm for glucose values)? I can only assume that G6 calibrations are handled by the transmitter directly and xDrip has no visibility/input into how they work. That would support why I don't see anything calibration-related on my extra status line, despite having "Calibration Data" selected to show. Some clarification on what applies to native G6 would be appreciated.

**Reply by Navid200** [reply]
In native mode, G6 transmitter handles the calibrations.  None of what is said here is of concern in that case:
https://navid200.github.io/xDrip/docs/Native-Algorithm.html

**Comment by Navid200** [comment]
This may be related:  https://github.com/NightscoutFoundation/xDrip/issues/318
